content/blog/eng/how-i-became-a-gitlab-contributor-to-fix-an-issue-in-our-projects-cicd-pipeline.md

```diff
@@ -0,0 +1,387 @@
+
+⚠️ Being this a workaround, however, there are some implications that need to be considered:
+
+* we can't define the `variables` as plain `YAML` properties as we did with **Docker Compose**
+* we still need to define a shared `variable` for each `service` at `job` level, meaning that `company-2-mock` can do `echo $COMPANY_1_ENVIRONMENT` and retrieve all the values for the other **mock server**, thus breaking the isolation between `services`
+* we need to manually add the original **Docker** `command` at the end of our overriden one, meaning that we need to keep it updated in case the base **Docker** image changes
+
+To overcome these issues we need a less "hacky" way to deal with our scenario, even if **GitLab** is not supporting it. Luckily **GitLab** is an **opensource** software, and this allows people to contribute to their code-base even if not directly employed by **GitLab** itself.
+
+As I said before I'm not a developer anymore, but I decided to go ahead and took my chance at adding the feature that we needed to improve our **CI/CD** pipelines.
+
+## Becoming a GitLab contributor to solve our issue with CI/CD
+
+>ℹ️
+>
+>To develop on **GitLab**'s platform you need the [GitLab Development Kit](https://gitlab.com/gitlab-org/gitlab-development-kit). The kit is well documented so I won't go into the details of the workflow.
+
+If you're not familiar with **GitLab**'s code-base, the initial experience will be daunting at best as their repository is quite huge. Lacking experience with **Ruby** didn't help either, but eventually I got a grip of which classes I needed to fiddle with in order to have my feature working.
+
+The feature proposal was to extend `services` definition by adding the same `variable` property that was already available for `jobs`, so that the validation logic and `variables` handling could be reused.
+
+This is the proposed structure, following the previous examples:
+
+```yaml
+job-1:
+    variables:
+        VAR: Job variable # This is meant as a variable to be used in job's script        
+    services:
+        - name: busybox:latest # Use a simple Busybox image so that we can a shell and just echo some variables
+          variables:
+            SVC_VAR: Service variable # This is meant as a variable to be used EXCLUSIVELY by the service
+          entrypoint: [ "sh", "-c", "echo $VAR, $SVC_VAR" ]
+    script:
+        - echo $VAR, $SVC_VAR
+```
+
+As you can see, `SVC_VAR` was now moved as a child of the `service` definition, meaning that it's not shared at the `job` level as before.
+
+Implementing this was easier than I thought, as I only had to copy the `variables` keyword from the `job` definitions (e.g. endpoints and internal data models) to their `service` counterparts. Between these changes, **GitLab**'s documentation and some **unit testing** I had only 20 added lines and 5 removed.
```

Comment:
Considering these changes, **GitLab** documentation, and some **unit testing** I only had 20 lines added and 5 removed.